README.rst
=========================================
Big Data Analytics Suite - Python Version
=========================================

This package implements three core algorithms
for big data analytics:
 * Principal Component Analysis (PCA)
 * K-Means Regression (Kmeans)
 * Support Vector Machine Classifier (SVM)

Each of the algorithms is packaged as a class
that accepts an N by M matrix of data points, X.
N is the number of points belonging to the local MPI
rank, and M is the number of features.
While N can vary between MPI ranks, M must be constant.

The SVM classifier also requires an array of N labels,
(for the local data points). The labels are
turned into 0 or 1 using the comparison (label > 0).


References:
 * [Refence Benchmark Results](https://arxiv.org/pdf/1811.02287.pdf)
 * [Coral2 Summary](https://asc.llnl.gov/sites/asc/files/2020-09/BDAS_Summary_b4bcf27_0.pdf)

Quick Start
===========

.. code-block::

    from bdas import SVM

    X, y = read_data("local_file.npy")
    S = SVM(X, y)
    for i in range(50):
        done = S.step(verb=True)
        print(f"Step {i} objective = {S.L}. Converged = {done}.")
        if done:
            break
    print(S.b)
    print(S.w)

When mpi4py support is enabled, you should
launch your program with `mpirun python3 my_prog.py`.

If you're using a supercomputer, consider installing
`spindle <https://computing.llnl.gov/projects/spindle/software>`_,
and then use `spindle mpirun python3 my_prog.py`.

---
SVM
---

The present implementation of SVM uses the maximum likelihood
estimation of the following observation model,

\[ P(x_i | w, b, z_i) =
  \frac{e^{z_i(w^T x_i - b)}}{1 + e^{w^T x_i - b}} 
\]
For simplicity, define
\[
s_i = \frac{e^{z_i(w^T x_i - b)}}{1 + e^{w^T x_i - b}}
.
\]

This model provides a sigmoidal probability of assigning
label $z_i$ to observation, $i$,
\[ P(z_i | w, b, x_i) =
  \frac{p_{z_i} e^{z_i(w^T x_i - b)}}{p_0 + p_1 e^{w^T x_i - b}}
\]
when $p_0$ and $p_1$ are the prior probabilities for
assigning categories 0 or 1 without knowing $x_i$.

Note that the distance from the separating plane to
point $x_i$ is $(w^T x_i - b)/|w| $.  That makes $ 1/|w| $
a scale parameter, which should properly have a scale-independent
(Jeffreys) prior,
\[
P(|w| | I) \propto |w|
\]

Combining the observation model with the prior,
taking the logarithm, and ignoring addititive constants
yields the posterior log-probability,
\[
L(b,w) = \log |w| + z^T (X w - b) - \sum_i \log(1 + e^{w^T x_i - b})
\]

This function is convex, and has a unique maximizer.

The interior-point optimization makes use of the
first and second derivatives of the objective,
(ordering variables b, then w).
\[
r(b,w) = \begin{bmatrix}
-1^T (z-s) \\
w/|w| ^2  + X^T (z-s)
\end{bmatrix}
\]
\[
J(b,w) = \begin{bmatrix}
-\mathrm{trace}(Q) & 1^T Q X \\
X^T Q 1 & I/|w| ^2 - 2w w^T/|w| ^4 - X^T Q X
\end{bmatrix}
\]
defining $Q = diag(s_i(1-s_i))$.


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Initial versions of 3 analysis methods.
  No implementation of parallel "allsum" primitives
  yet.


pca.py
#!/usr/bin/env python3
import numpy as np
import numpy.random as random
import numpy.linalg as la

class PCA:
    def __init__(self, X):
        self.N = X.shape[0]
        self.M = X.shape[1]
        self.Ntot = self.allsum(self.N)
        # mean of the data
        self.m = self.allsum( X.sum(0) ) / self.Ntot

        # variance-covariance matrix
        C = np.dot((X-self.m).T, (X-self.m))
        self.C = self.allsum(C) / self.Ntot
        val, vec = la.eigh(self.C)

        # scale of each direction
        self.scale = np.sqrt(val)
        # variance-covariance directions
        self.vec = vec
    def allsum(self, x):
        return x

def randrot(i,j,N):
    assert i >= 0 and j >= 0 and i < N and j < N
    assert i != j
    c = random.uniform()
    s = np.sqrt(1.0-c*c)
    R = np.identity(N)
    R[i,i] = c
    R[i,j] = s
    R[j,i] = -s
    R[j,j] = c
    return R

def test1(N, M):
    # Cholesky decompose of covariance matrix
    # positive eigenvalues uniformly in [10,90]
    val = random.random(M)*90.0 + 10.0
    L = np.diag( val )
    val = np.sort(val)
    print(val)
    for i in range(M):
        for j in range(i+1,M):
            L = np.dot(randrot(i,j,M), L)

    # local samples
    X = np.dot(random.standard_normal((N,M)), L.T)
    S = PCA(X)
    # These 2 should be similar
    if M <= 5:
        print(S.C)
        print(np.dot(L, L.T))
    print(S.scale)

if __name__=="__main__":
    test1(1000, 5)
    #test1(1000, 100)


blob.py
# Example SVM solution plotting code from scikit-learn
# modified to show results of bayes.SVM.
#
import numpy as np
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn.datasets import make_blobs
from bayes import SVM

# we create 40 separable points
X, y = make_blobs(n_samples=40, centers=2, random_state=6)
print(X.shape)

# fit the model, don't regularize for illustration purposes
#clf = svm.SVC(kernel='linear', C=1000)
#clf.fit(X, y)
print(y)
clf = SVM(X, y)
for i in range(50):
    done = clf.step(verb=True)
    print(f"Step {i} objective = {clf.L}. Converged = {done}.")
    if done:
        break

plt.scatter(X[:, 0], X[:, 1], c=y, s=30, cmap=plt.cm.Paired)

# plot the decision function
ax = plt.gca()
xlim = ax.get_xlim()
ylim = ax.get_ylim()

# create grid to evaluate model
xx = np.linspace(xlim[0], xlim[1], 60)
yy = np.linspace(ylim[0], ylim[1], 60)
YY, XX = np.meshgrid(yy, xx)
xy = np.vstack([XX.ravel(), YY.ravel()]).T
Z = clf.decision_function(xy).reshape(XX.shape)

# plot decision boundary and margins
ax.contour(XX, YY, Z, colors='k', levels=[-1, 0, 1], alpha=0.5,
           linestyles=['--', '-', '--'])
# plot support vectors
#ax.scatter(clf.support_vectors_[:, 0], clf.support_vectors_[:, 1], s=100,
#           linewidth=1, facecolors='none', edgecolors='k')
plt.show()


kmeans.py
#!/usr/bin/env python3

import numpy as np
import numpy.linalg as la
import numpy.random as random

# FIXME: this routine needs to distribute the
# choice of K centers among processors.
def choose_K(X, K, Ntot):
    #if C.rank == 0:
    #else:
    #    m
    ind = random.choice(Ntot, size=K, replace=False)
    return X[ind]

class Kmeans:
    def __init__(self, X, K):
        self.X = X
        self.K = K
        self.N = X.shape[0]
        self.Ntot = self.allsum(self.N)
        self.M = X.shape[1]
        self.z = None # category labels
        self.occ = None # total in each category
        self.ctr = choose_K(self.X, K, self.Ntot)
        # total L2 distance within each category
        # (filled in by categorize)
        self.err = None

    def allsum(self, x):
        return x

    def categorize(self, calc_err=True):
        dist = self.X[:,None,:] - self.ctr[None,:,:]
        d2 = np.sum(dist*dist,2)
        self.z = np.argmin(d2, 1)
        if calc_err:
            self.occ = np.zeros(self.K, int)
            self.err = np.zeros(self.K)
            for k in range(self.K):
                self.occ[k] = np.sum(self.z == k)
                self.err[k] = np.sum(d2[self.z == k]**0.5)
            self.occ = self.allsum(self.occ)
            self.err = self.allsum(self.err)

    def estimate(self):
        # column 0 are the occupancies
        loc_ctr = np.zeros((self.K,1+self.M))
        for k in range(self.K):
            m = self.z == k
            loc_ctr[k,1:] = np.dot(m, self.X)
            loc_ctr[k,0] = np.sum(m)
        loc_ctr = self.allsum(loc_ctr)

        self.occ = loc_ctr[:,0]
        self.ctr = loc_ctr[:,1:]/loc_ctr[:,0,None]

    def step(self):
        # On the first iteration, determines labels and exits.
        #
        # For all other iterations, assumes categories (aka labels)
        # are already determined, and does:
        #  1. adjust centroids
        #  2. re-label
        #
        # Returns True when converged within tol
        if self.err is None:
            self.categorize(True)
            return False
        z0 = self.z.copy()
        err = self.err.sum()
        
        self.estimate()
        self.categorize(True)
        err2 = self.err.sum()
        # how many categories changed?
        dz = self.allsum( np.sum(self.z != z0) )
        #if np.abs(err - err2) < tol:
        #    return True
        return dz == 0

    def show(S):
        print(f"Ctr = {S.ctr}")
        print(f"categories = {S.z}")
        if S.occ is not None:
            print(f"Cluster occupancy: {S.occ}")
        if S.err is not None:
            print(f"Average distance to centers: {S.err/(S.occ+(S.occ == 0))}")

def test1(N, M, K):
    sigma = 0.6
    ctrs = np.array([[0.0, 0.0, 0.0],
                     [1.0, 1.0, 1.0],
                     [-1.0, 1.0, 1.0],
                     [-1.0, -0.5, -0.5]])
    assert K <= ctrs.shape[0]
    assert M <= ctrs.shape[1]
    assert N >= K
    z = np.zeros(N)
    X = np.zeros((N,M))
    for i in range(K):
        d = ((i+1)*N)//K - (i*N)//K
        z[(i*N)//K : ((i+1)*N)//K] = i
        X[(i*N)//K : ((i+1)*N)//K] = sigma*random.standard_normal((d,M)) \
                                   + ctrs[i,:M]

    S = Kmeans(X, K)
    for i in range(10):
        done = S.step()
        if done:
            print(f"Converged in {i} steps.")
            break
    else:
        print("Convergence not obtained.")
    S.show()

if __name__=="__main__":
    test1(10, 2, 3)
    test1(100, 3, 3)
    test1(1000, 3, 4)


bayes.py
#!/usr/bin/env python3
# Bayesian SVM regression

import numpy as np
import numpy.linalg as la
import numpy.random as random

# Asymptotics:
#   as x ->  infty, softplus(x) -> x      [infty]
#   as x -> -infty, softplus(x) -> exp(x) [0]
def softplus(x):
    y = x.copy()
    m = np.where(y < 10.0)
    if len(m[0]) > 0:
        y[m] = np.log1p(np.exp(x[m]))
    return y

class SVM:
    # X is the (samples x features) matrix of data
    # and z is the categorization vector (len(z)=samples).
    # z[i] should be 0 or 1, depending on the category of i.
    #
    # This class can be called in parallel.
    # In this case, it holds only local rows of X, z.
    # and (intermediate variable) s.
    # Everything else is global.
    def __init__(self, X, z, alpha=0.2, beta=0.8):
        # problem instance
        self.X = X
        self.z = z > 0 # turn into 2-category information
        self.N = X.shape[0]
        self.M = X.shape[1]
        self.Ntot = self.allsum(self.N)

        # solution variables (updated during step)
        # start off pointing at data centroid
        c1 = self.allsum(np.dot(self.z, self.X))/self.Ntot # z1 centroid
        c0 = self.allsum(np.sum(self.X,0))/self.Ntot # data centroid
        c1 = 2*c1-c0 # hyperplane direction
        self.w = c1 / np.dot(c1,c1)**0.5
        self.b = np.dot(self.w, c0)

        # algorithm parameters affecting line-search:
        assert 0.0 < alpha and alpha < 0.5
        assert 0.0 < beta and beta < 1.0
        self.alpha = alpha
        self.beta = beta

        self.Dval = None

        N1 = self.allsum(self.z.sum())
        N0 = self.Ntot - N1
        print(N0, N1)
        # constant shift to log-posterior likelihood
        self.log_const = N0*np.log(N0 + (N0 == 0)) \
                       + N1*np.log(N1 + (N1 == 0)) \
                       - self.Ntot*np.log(self.Ntot)
        self.N0 = N0
        self.N1 = N1

    @property
    def L(self):
        # Report the log-likelihood in per-sample units.
        if self.Dval is None:
            return None
        ans = (self.Dval+self.log_const) / self.Ntot
        #print(ans)
        return ans

    def D(self, w, p):
        # Return the current value of the objective function
        #
        ans = np.dot(p, self.z) \
            - softplus(p).sum()
        ans = self.allsum(ans)

        return ans + 0.5*np.log(np.dot(w,w))

    def calc_p(self, b, w):
        return np.dot(self.X, w) - b

    def calc_ps(self):
        p = self.calc_p(self.b, self.w)
        t = p.clip(None, 0.0)
        s = np.exp(t) / (1.0 + np.exp(2*t-p))
        return p, s

    def step(self, rtol=1e-6, Ltol=1e-9, verb=False):
        # Take a single step of unconstrained
        # Newton's method with backtracking.
        #
        # On output, updates self.b and self.w.
        # Also caches the value of D in self.Dval.
        # Returns True if minimization is complete,
        # and False if not.
        # Completion is achieved if |r| < rtol
        # or |L-L'| < Ltol.
        #
        p, s = self.calc_ps()

        if self.Dval is None:
            self.Dval = self.D(self.w, p)
            if verb:
                print(f"Initial objective = {self.L}")

        # Note: computing r and J each do 1 allsum,
        # so they be combined.
        r = self.r(s) # residual
        # check convergence based on |r|
        rnorm2 = self.allsum(np.vdot(r, r))
        if verb:
            print(f"Residual norm = {rnorm2**0.5}")
        if rnorm2 < rtol**2:
            return True
        # search direction
        d = -la.solve(self.calc_J(s), r)

        # linear extrapolation, D(t) ~ D(0) + t*slope/alpha
        slope = self.alpha*np.dot(r,d)
        if verb:
            # scale the slope to input units
            scale = self.Ntot*self.alpha
            print(f"Residual slope along d = {slope/scale}")
        t = 1.0 # full step
        while True:
            b = self.b + t*d[0]
            w = self.w + t*d[1:]
            D = self.D(w, self.calc_p(b, w))
            if D >= self.Dval + t*slope:
                break
            t = t*self.beta # scale down step
        if verb and t != 1.0:
            print(f"Back-tracked to t = {t}")

        # check change in objective
        done = np.abs(D-self.Dval) < Ltol*self.Ntot
        self.Dval = D
        self.b = b
        self.w = w
        return done

    def allsum(self, x):
        return x

    def r(self, s):
        r = np.zeros(1+self.M)
        r[0] = np.sum(s)
        r[1:] = np.dot(self.z-s, self.X)
        r = self.allsum(r)

        w2 = np.dot(self.w, self.w)
        r[0] -= self.N1 # sum z
        r[1:] += self.w/w2
        return r

    def calc_J(self, s):
        J = np.zeros((1+self.M,1+self.M))
        Q = s*(1.0-s)
        J[0,0] = -Q.sum()
        J[0,1:] = np.dot(Q,self.X)
        J[1:,0] = J[0,1:]
        J[1:,1:] = -np.dot(self.X.T*Q, self.X)

        w2 = np.dot(self.w, self.w)
        J = self.allsum(J)
        J[1:,1:] += (1.0/w2) * np.identity(self.M) \
                - (2.0*w2**-2) * self.w[:,None]*self.w[None,:]

        #if np.vdot(J[0],J[0]) < 1e-8:
        #    J[0,0] = 0.001

        return J

    def decision_function(self, X, tol=0.2):
        assert tol > 0.0 and tol <= 0.5
        print(self.b, self.w)
        p = np.dot(X, self.w) - self.b
        a = np.log(tol/(1.0-tol))
        b = np.log((1.0-tol)/tol)
        print(a, b)
        return p/b
        # clamp ranges
        #p[p < a] = -1.0
        #p[p >= b] = 1.0
        #p[(p >= b)*(p < a)] = 0.0
        #return p

def Ndiff(f, b, x0):
    h = 1e-7
    ih = 1.0/h
    h = 1.0/ih
    x1 = x0.copy()
    f0 = f(b, x0)
    f1 = np.zeros((1+len(x0),)+f0.shape)
    f1[0] = f(b+h,x0)
    for i in range(len(x0)):
        x1[i] = x0[i]+h
        f1[i+1] = f(b,x1)
        x1[i] = x0[i]
    return ih*(f1-f0)

def test1(N=10, M=4):
    y = np.arange(N) < N/2 # 0 or 1 for now
    x = random.standard_normal((N,M)) + 1.0 + y[:,None]
    y = 2*y-1 # -1 or 1
    S = SVM(x, y)
    for i in range(10):
        done = S.step(verb=True)
        if i == 2:
            check_diff(S)
        print(f"Step {i} objective = {S.L}. Converged = {done}.")
        if done:
            break

    print(S.b)
    print(S.w)

def check_diff(S):
    b0 = S.b
    w0 = S.w
    def calcD(b,w):
        return S.D(w, S.calc_p(b,w))
    def calcr(b,w):
        S.b = b
        S.w = w
        p, s = S.calc_ps()
        return S.r(s)
    
    r1 = Ndiff(calcD, b0, w0)
    r0 = calcr(b0, w0)
    err = np.abs(r1-r0).max()
    print(f"Residual err = {err}")
    if err > 1e-4:
        print(r0)
        print(r1)

    J1 = Ndiff(calcr, b0, w0)
    S.b = b0
    S.w = w0
    p, s = S.calc_ps()
    J0 = S.calc_J(s)
    err = np.abs(J1-J0).max()
    print(f"J err = {err}")
    if err > 1e-4:
        print(J0)
        print(J1)

def read_data(fname):
    # reads a data file with raw lines like:
    # -1 3 11 14 19 39 42 55 64 67 73 75 76 80 83 
    x = []
    y = []
    with open(fname, encoding='utf-8') as f:
        for line in f:
            tok = line.split()
            if len(tok) < 15:
                continue
            x.append([float(t) for t in tok[1:]])
            y.append(float(tok[0]))
    return np.array(x), np.array(y)

def test2(fname):
    x, y = read_data(fname)
    print(x.shape, y.shape)
    S = SVM(x, y)
    for i in range(50):
        done = S.step(verb=True)
        print(f"Step {i} objective = {S.L}. Converged = {done}.")
        if done:
            break
    print(S.b)
    print(S.w)

if __name__=="__main__":
    test1(1000, 2)
    #test2("a1a.t")
